File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/EnterPinPresenter.java
package dev.ragnarok.fenrir.mvp.presenter;

import static dev.ragnarok.fenrir.util.Utils.isEmpty;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Arrays;
import java.util.List;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.domain.IOwnersRepository;
import dev.ragnarok.fenrir.domain.Repository;
import dev.ragnarok.fenrir.model.Owner;
import dev.ragnarok.fenrir.mvp.presenter.base.RxSupportPresenter;
import dev.ragnarok.fenrir.mvp.view.IEnterPinView;
import dev.ragnarok.fenrir.settings.ISettings;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.Objects;
import dev.ragnarok.fenrir.util.RxUtils;

public class EnterPinPresenter extends RxSupportPresenter<IEnterPinView> {

    private static final String SAVE_VALUE = "save_value";
    private static final int LAST_CIRCLE_VISIBILITY_DELAY = 200;
    private static final int NO_VALUE = -1;
    private static final int MAX_ATTEMPT_DELAY = 3 * 60 * 1000;
    private final IOwnersRepository ownersRepository;
    private final ISettings.ISecuritySettings securitySettings;
    private final int[] mValues;
    private final Handler mHandler = new Handler(Looper.getMainLooper());
    private final Runnable mOnFullyEnteredRunnable = this::onFullyEntered;
    private boolean onShowedFirstFingerPrint;
    private Owner mOwner;

    public EnterPinPresenter(@Nullable Bundle savedState) {
        super(savedState);
        securitySettings = Settings.get().security();
        ownersRepository = Repository.INSTANCE.getOwners();

        if (savedState != null) {
            mValues = savedState.getIntArray(SAVE_VALUE);
        } else {
            mValues = new int[Constants.PIN_DIGITS_COUNT];
            resetPin();
        }

        if (Objects.isNull(mOwner)) {
            loadOwnerInfo();
        }
    }

    private void loadOwnerInfo() {
        int accountId = Settings.get()
                .accounts()
                .getCurrent();

        if (accountId != ISettings.IAccountsSettings.INVALID_ID) {
            appendDisposable(ownersRepository.getBaseOwnerInfo(accountId, accountId, IOwnersRepository.MODE_ANY)
                    .compose(RxUtils.applySingleIOToMainSchedulers())
                    .subscribe(this::onOwnerInfoReceived, t -> {/*ignore*/}));
        }
    }

    private void onOwnerInfoReceived(Owner owner) {
        mOwner = owner;
        resolveAvatarView();
    }

    public void onFingerprintClicked() {
        if (!securitySettings.isEntranceByFingerprintAllowed()) {
            callView(v -> v.getCustomToast().showToastError(R.string.error_login_by_fingerprint_not_allowed));
            return;
        }
        callView(IEnterPinView::showBiometricPrompt);
    }

    private void resolveAvatarView() {
        String avatar = Objects.isNull(mOwner) ? null : mOwner.getMaxSquareAvatar();
        if (isEmpty(avatar)) {
            callView(IEnterPinView::displayDefaultAvatar);
        } else {
            callView(v -> v.displayAvatarFromUrl(avatar));
        }
    }

    private long getNextPinAttemptTimeout() {
        List<Long> history = Settings.get()
                .security()
                .getPinEnterHistory();

        if (history.size() < Settings.get().security().getPinHistoryDepth()) {
            return 0;
        }

        long howLongAgoWasFirstAttempt = System.currentTimeMillis() - history.get(0);
        return howLongAgoWasFirstAttempt < MAX_ATTEMPT_DELAY ? MAX_ATTEMPT_DELAY - howLongAgoWasFirstAttempt : 0;
    }

    private void refreshViewCirclesVisibility() {
        callView(v -> v.displayPin(mValues, NO_VALUE));
    }

    public void onBackspaceClicked() {
        int currentIndex = getCurrentIndex();
        if (currentIndex == -1) {
            mValues[mValues.length - 1] = NO_VALUE;
        } else if (currentIndex > 0) {
            mValues[currentIndex - 1] = NO_VALUE;
        }

        refreshViewCirclesVisibility();
    }

    private void onFullyEntered() {
        if (!isFullyEntered()) return;

        long timeout = getNextPinAttemptTimeout();
        if (timeout > 0) {
            callView(view -> view.showError(R.string.limit_exceeded_number_of_attempts_message, timeout / 1000));

            resetPin();
            refreshViewCirclesVisibility();
            return;
        }

        Settings.get()
                .security()
                .firePinAttemptNow();

        if (Settings.get().security().isPinValid(mValues)) {
            onEnteredRightPin();
        } else {
            onEnteredWrongPin();
        }
    }

    private void onEnteredRightPin() {
        Settings.get()
                .security()
                .clearPinHistory();

        callView(IEnterPinView::sendSuccessAndClose);
    }

    private void onEnteredWrongPin() {
        resetPin();
        refreshViewCirclesVisibility();

        callView(v -> {
            v.showError(R.string.pin_is_invalid_message);
            v.displayErrorAnimation();
        });
    }

    public void onNumberClicked(int value) {
        if (isFullyEntered()) return;

        mValues[getCurrentIndex()] = value;
        refreshViewCirclesVisibility();

        if (isFullyEntered()) {
            mHandler.removeCallbacks(mOnFullyEnteredRunnable);
            mHandler.postDelayed(mOnFullyEnteredRunnable, LAST_CIRCLE_VISIBILITY_DELAY);
        }
    }

    private boolean isFullyEntered() {
        for (int value : mValues) {
            if (value == NO_VALUE) {
                return false;
            }
        }

        return true;
    }

    private int getCurrentIndex() {
        for (int i = 0; i < mValues.length; i++) {
            if (mValues[i] == NO_VALUE) {
                return i;
            }
        }

        return -1;
    }

    @Override
    public void onGuiCreated(@NonNull IEnterPinView view) {
        super.onGuiCreated(view);
        if (!onShowedFirstFingerPrint) {
            onShowedFirstFingerPrint = true;
            if (securitySettings.isEntranceByFingerprintAllowed()) {
                view.showBiometricPrompt();
            }
        }
        resolveAvatarView();
        refreshViewCirclesVisibility();
    }

    private void resetPin() {
        Arrays.fill(mValues, NO_VALUE);
    }

    @Override
    public void saveState(@NonNull Bundle outState) {
        super.saveState(outState);
        outState.putIntArray(SAVE_VALUE, mValues);
    }
}

File: app_filegallery/src/main/kotlin/dev/ragnarok/filegallery/mvp/presenter/EnterPinPresenter.kt
package dev.ragnarok.filegallery.mvp.presenter

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import dev.ragnarok.filegallery.Constants
import dev.ragnarok.filegallery.R
import dev.ragnarok.filegallery.mvp.presenter.base.RxSupportPresenter
import dev.ragnarok.filegallery.mvp.view.IEnterPinView
import dev.ragnarok.filegallery.settings.Settings
import java.util.*

class EnterPinPresenter(savedState: Bundle?) : RxSupportPresenter<IEnterPinView>(savedState) {
    private val mValues: IntArray?
    private val mHandler = Handler(Looper.getMainLooper())
    private val mOnFullyEnteredRunnable = Runnable { onFullyEntered() }
    private var onShowedFirstFingerPrint = false

    fun onFingerprintClicked() {
        if (!Settings.get().security().isEntranceByFingerprintAllowed) {
            view?.customToast?.showToastError(
                R.string.error_login_by_fingerprint_not_allowed
            )
            return
        }
        view?.showBiometricPrompt()
    }

    private val nextPinAttemptTimeout: Long
        get() {
            val history = Settings.get()
                .security()
                .pinEnterHistory
            if (history.size < Settings.get().security().pinHistoryDepthValue()) {
                return 0
            }
            val howLongAgoWasFirstAttempt = System.currentTimeMillis() - history[0]
            return if (howLongAgoWasFirstAttempt < MAX_ATTEMPT_DELAY) MAX_ATTEMPT_DELAY - howLongAgoWasFirstAttempt else 0
        }

    private fun refreshViewCirclesVisibility() {
        if (mValues != null) {
            view?.displayPin(
                mValues,
                NO_VALUE
            )
        }
    }

    fun onBackspaceClicked() {
        val currentIndex = currentIndex
        if (currentIndex == -1) {
            mValues?.set(mValues.size - 1, NO_VALUE)
        } else if (currentIndex > 0) {
            mValues?.set(currentIndex - 1, NO_VALUE)
        }
        refreshViewCirclesVisibility()
    }

    private fun onFullyEntered() {
        if (!isFullyEntered) return
        val timeout = nextPinAttemptTimeout
        if (timeout > 0) {
            view?.showError(
                R.string.limit_exceeded_number_of_attempts_message,
                timeout / 1000
            )
            resetPin()
            refreshViewCirclesVisibility()
            return
        }
        Settings.get()
            .security()
            .firePinAttemptNow()
        if (mValues?.let { Settings.get().security().isPinValid(it) } == true) {
            onEnteredRightPin()
        } else {
            onEnteredWrongPin()
        }
    }

    private fun onEnteredRightPin() {
        Settings.get()
            .security()
            .clearPinHistory()
        view?.sendSuccessAndClose()
    }

    private fun onEnteredWrongPin() {
        resetPin()
        refreshViewCirclesVisibility()
        view?.let {
            it.showError(R.string.pin_is_invalid_message)
            it.displayErrorAnimation()
        }
    }

    fun onNumberClicked(value: Int) {
        if (isFullyEntered) return
        mValues?.set(currentIndex, value)
        refreshViewCirclesVisibility()
        if (isFullyEntered) {
            mHandler.removeCallbacks(mOnFullyEnteredRunnable)
            mHandler.postDelayed(mOnFullyEnteredRunnable, LAST_CIRCLE_VISIBILITY_DELAY.toLong())
        }
    }

    private val isFullyEntered: Boolean
        get() {
            if (mValues != null) {
                for (value in mValues) {
                    if (value == NO_VALUE) {
                        return false
                    }
                }
            }
            return true
        }
    private val currentIndex: Int
        get() {
            mValues?.let {
                for (i in mValues.indices) {
                    if (mValues[i] == NO_VALUE) {
                        return i
                    }
                }
            }
            return -1
        }

    override fun onGuiCreated(viewHost: IEnterPinView) {
        super.onGuiCreated(viewHost)
        if (!onShowedFirstFingerPrint) {
            onShowedFirstFingerPrint = true
            if (Settings.get().security().isEntranceByFingerprintAllowed) {
                viewHost.showBiometricPrompt()
            }
        }
        refreshViewCirclesVisibility()
    }

    private fun resetPin() {
        if (mValues != null) {
            Arrays.fill(mValues, NO_VALUE)
        }
    }

    override fun saveState(outState: Bundle) {
        super.saveState(outState)
        outState.putIntArray(SAVE_VALUE, mValues)
    }

    companion object {
        private const val SAVE_VALUE = "save_value"
        private const val LAST_CIRCLE_VISIBILITY_DELAY = 200
        private const val NO_VALUE = -1
        private const val MAX_ATTEMPT_DELAY = 3 * 60 * 1000
    }

    init {
        if (savedState != null) {
            mValues = savedState.getIntArray(SAVE_VALUE)
        } else {
            mValues = IntArray(Constants.PIN_DIGITS_COUNT)
            resetPin()
        }
    }
}